File: lib/ai.js
// Optional vision review of the captured screenshots.
//
// Provider is auto-detected from whichever API key is present, preferring the
// FREE options first so no billing is required:
//   1. Google Gemini   — GEMINI_API_KEY (or GOOGLE_API_KEY). FREE tier at
//                         https://aistudio.google.com/apikey (no credit card).
//   2. Groq            — GROQ_API_KEY. Free tier, OpenAI-compatible.
//   3. OpenAI          — OPENAI_API_KEY. Paid.
//
// All calls use plain fetch (Node 20 global) — no SDK dependency.

const SYSTEM = [
  'You are a meticulous QA engineer reviewing screenshots of a freshly deployed web app.',
  'You are given screenshots at one or more viewport sizes, a short digest of the page\'s visible text,',
  'and a list of specifics the owner wants watched for. Judge ONLY from the evidence provided.',
  'Be concrete and specific; cite what you actually see. Do not invent problems that are not visible.',
  'Respond with a single JSON object matching the requested schema.',
].join(' ');

const SCHEMA_HINT = `Return JSON exactly in this shape:
{
  "overall_impression": "one or two sentences on how the page looks",
  "looks_broken": true|false,
  "visual_bugs": [ { "severity": "high|medium|low", "issue": "...", "where": "which viewport/area" } ],
  "feature_verdicts": [ { "feature": "<echo the specific>", "status": "present|partial|missing|broken", "evidence": "what in the screenshot supports this" } ]
}
If the specifics list is empty, return an empty feature_verdicts array. Output only the JSON, no prose.`;

function geminiKey() { return process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY; }

function pickProvider() {
  if (geminiKey()) return 'gemini';
  if (process.env.GROQ_API_KEY) return 'groq';
  if (process.env.OPENAI_API_KEY) return 'openai';
  return null;
}

export function aiConfigured() {
  return !!pickProvider();
}

export async function aiReview(audit, features) {
  const provider = pickProvider();
  if (!provider) {
    return {
      available: false,
      reason: 'No AI key set. Add a FREE Gemini key as GEMINI_API_KEY ' +
        '(get one at https://aistudio.google.com/apikey — no billing needed).',
    };
  }

  const shots = (audit.screenshots || []).filter((s) => s.b64).slice(0, 3);
  if (!shots.length) return { available: false, reason: 'No screenshots available to analyze.' };

  const digest = (audit.page_text || '').slice(0, 1500);
  const featTxt = (features || []).filter(Boolean).map((f) => `- ${f}`).join('\n') || '(none provided)';
  const userText = `URL under test: ${audit.url}\n\nVisible-text digest:\n${digest}\n\n` +
    `Specifics the owner wants watched for:\n${featTxt}\n\n${SCHEMA_HINT}`;

  try {
    if (provider === 'gemini') return await runGemini(shots, userText);
    return await runOpenAICompatible(provider, shots, userText);
  } catch (e) {
    return { available: false, reason: `${e.name || 'Error'}: ${e.message || e}` };
  }
}

// ---- Google Gemini (free) ---------------------------------------------------
async function runGemini(shots, userText) {
  const key = geminiKey();
  const parts = [];
  for (const s of shots) {
    parts.push({ text: `Screenshot — ${s.name} (${s.width}x${s.height}):` });
    parts.push({ inline_data: { mime_type: 'image/png', data: s.b64 } });
  }
  parts.push({ text: userText });
  const body = JSON.stringify({
    system_instruction: { parts: [{ text: SYSTEM }] },
    contents: [{ role: 'user', parts }],
    generationConfig: { responseMimeType: 'application/json', maxOutputTokens: 1600, temperature: 0.2 },
  });

  // Try a sequence of free-tier models — the newest ones can 429 on a fresh key,
  // while the lighter flash models have generous free quota. Respect an explicit
  // GEMINI_MODEL override.
  const models = process.env.GEMINI_MODEL
    ? [process.env.GEMINI_MODEL]
    : ['gemini-2.0-flash', 'gemini-2.0-flash-lite', 'gemini-1.5-flash', 'gemini-1.5-flash-8b'];

  let lastErr = '';
  for (const model of models) {
    const url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${encodeURIComponent(key)}`;
    let resp;
    try {
      resp = await fetch(url, { method: 'POST', headers: { 'Content-Type': 'application/json' }, body });
    } catch (e) {
      lastErr = `${model}: ${e.message || e}`;
      continue;
    }
    if (resp.ok) {
      const json = await resp.json();
      const raw = (json.candidates?.[0]?.content?.parts || []).map((p) => p.text || '').join('').trim();
      const data = safeParse(raw);
      if (!data) return { available: true, model, error: 'Could not parse model response.', raw: raw.slice(0, 600) };
      return { available: true, model: `${model} (Gemini, free)`, ...data };
    }
    const bodyText = await resp.text().catch(() => '');
    lastErr = `${model} → ${resp.status}`;
    // A bad/blocked key won't be fixed by trying another model.
    if (resp.status === 401 || resp.status === 403) {
      return { available: false, reason: `Gemini rejected the key (${resp.status}). Check GEMINI_API_KEY.` };
    }
    // 429 (quota) / 404 (model not available) → try the next model.
  }
  return {
    available: false,
    reason: `Gemini free quota/model unavailable (last: ${lastErr}). ` +
      `The free quota may be temporarily used up for this key — retry shortly, ` +
      `or add a free GROQ_API_KEY instead.`,
  };
}

// ---- OpenAI-compatible (Groq free, or OpenAI paid) --------------------------
async function runOpenAICompatible(provider, shots, userText) {
  const cfg = provider === 'groq'
    ? {
        key: process.env.GROQ_API_KEY,
        base: 'https://api.groq.com/openai/v1',
        model: process.env.GROQ_MODEL || 'meta-llama/llama-4-scout-17b-16e-instruct',
        label: 'Groq, free',
      }
    : {
        key: process.env.OPENAI_API_KEY,
        base: 'https://api.openai.com/v1',
        model: process.env.OPENAI_MODEL || 'gpt-4o',
        label: 'OpenAI',
      };

  const content = [];
  for (const s of shots) {
    content.push({ type: 'text', text: `Screenshot — ${s.name} (${s.width}x${s.height}):` });
    content.push({ type: 'image_url', image_url: { url: `data:image/png;base64,${s.b64}` } });
  }
  content.push({ type: 'text', text: userText });

  const resp = await fetch(`${cfg.base}/chat/completions`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${cfg.key}` },
    body: JSON.stringify({
      model: cfg.model,
      max_tokens: 1400,
      temperature: 0.2,
      response_format: { type: 'json_object' },
      messages: [{ role: 'system', content: SYSTEM }, { role: 'user', content }],
    }),
  });
  if (!resp.ok) {
    const body = await resp.text().catch(() => '');
    return { available: false, reason: `${cfg.label} API ${resp.status}: ${body.slice(0, 200)}` };
  }
  const json = await resp.json();
  const raw = json.choices?.[0]?.message?.content || '';
  const data = safeParse(raw);
  if (!data) return { available: true, model: cfg.model, error: 'Could not parse model response.', raw: raw.slice(0, 600) };
  return { available: true, model: `${cfg.model} (${cfg.label})`, ...data };
}

function safeParse(raw) {
  raw = String(raw || '').trim();
  if (raw.startsWith('```')) raw = raw.replace(/^```(?:json)?/i, '').replace(/```$/, '').trim();
  try { return JSON.parse(raw); } catch (e) { /* fall through */ }
  const a = raw.indexOf('{'), b = raw.lastIndexOf('}');
  if (a >= 0 && b > a) { try { return JSON.parse(raw.slice(a, b + 1)); } catch (e) { /* noop */ } }
  return null;
}

File: vercel_bot/ai.py
"""
Optional Claude-vision layer.

Given the captured screenshots, a short DOM text digest and the user's list of
expected features, ask Claude to (a) spot visual bugs a DOM audit can't see and
(b) give a per-feature verdict grounded in the screenshots.

Everything here is best-effort and optional: if no API key is configured, or the
call fails, the caller simply skips the AI section. The heuristic engine is the
source of truth; the AI is an enhancement.
"""

from __future__ import annotations

import base64
import json
import os
from typing import Any, Dict, List, Optional

# A capable, fast vision model. Overridable via env for cost/latency tuning.
DEFAULT_MODEL = os.environ.get("VERCELCHECK_AI_MODEL", "claude-sonnet-5")

_SYSTEM = (
    "You are a meticulous QA engineer reviewing screenshots of a freshly deployed "
    "web app. You are given screenshots at one or more viewport sizes, a short "
    "digest of the page's visible text, and a list of features/content the owner "
    "says the app SHOULD include. "
    "Judge only from the evidence provided. Be concrete and specific; cite what you "
    "see. Do not invent problems that aren't visible. Respond with a single JSON "
    "object and nothing else."
)

_SCHEMA_HINT = """
Return JSON exactly in this shape:
{
  "overall_impression": "one or two sentences on how the page looks",
  "looks_broken": true|false,
  "visual_bugs": [
    {"severity": "high|medium|low", "issue": "...", "where": "which viewport/area"}
  ],
  "feature_verdicts": [
    {"feature": "<echo the feature>", "status": "present|partial|missing|broken",
     "evidence": "what in the screenshot supports this"}
  ]
}
If a features list is empty, return an empty feature_verdicts array.
""".strip()


def is_available(api_key: Optional[str] = None) -> bool:
    key = api_key or os.environ.get("ANTHROPIC_API_KEY")
    if not key:
        return False
    try:
        import anthropic  # noqa: F401
    except Exception:
        return False
    return True


def _img_block(path: str) -> Optional[Dict[str, Any]]:
    try:
        with open(path, "rb") as fh:
            data = fh.read()
    except Exception:
        return None
    if not data:
        return None
    return {
        "type": "image",
        "source": {
            "type": "base64",
            "media_type": "image/png",
            "data": base64.standard_b64encode(data).decode("ascii"),
        },
    }


def analyze(audit: Dict[str, Any], features: List[str],
            api_key: Optional[str] = None,
            model: str = DEFAULT_MODEL) -> Dict[str, Any]:
    """Run the vision analysis. Returns {"available": bool, ...}."""
    key = api_key or os.environ.get("ANTHROPIC_API_KEY")
    if not key:
        return {"available": False, "reason": "No ANTHROPIC_API_KEY configured."}
    try:
        import anthropic
    except Exception:
        return {"available": False, "reason": "anthropic package not installed."}

    content: List[Dict[str, Any]] = []
    # Attach up to 3 screenshots, labeled.
    for shot in audit.get("screenshots", [])[:3]:
        img = _img_block(shot["path"])
        if img:
            content.append({"type": "text",
                            "text": f"Screenshot — {shot['name']} ({shot['width']}x{shot['height']}):"})
            content.append(img)

    if not any(b.get("type") == "image" for b in content):
        return {"available": False, "reason": "No screenshots available to analyze."}

    digest = (audit.get("page_text") or "")[:1500]
    feat_txt = "\n".join(f"- {f}" for f in features if f.strip()) or "(none provided)"
    content.append({
        "type": "text",
        "text": (
            f"URL under test: {audit.get('url')}\n\n"
            f"Visible-text digest:\n{digest}\n\n"
            f"Features the owner says it SHOULD include:\n{feat_txt}\n\n"
            f"{_SCHEMA_HINT}"
        ),
    })

    try:
        client = anthropic.Anthropic(api_key=key)
        msg = client.messages.create(
            model=model,
            max_tokens=1600,
            system=_SYSTEM,
            messages=[{"role": "user", "content": content}],
        )
        raw = "".join(b.text for b in msg.content if getattr(b, "type", "") == "text").strip()
        data = _parse_json(raw)
        if data is None:
            return {"available": True, "error": "Could not parse model response.",
                    "raw": raw[:800]}
        data["available"] = True
        data["model"] = model
        return data
    except Exception as e:  # network / auth / rate-limit
        return {"available": False, "reason": f"{type(e).__name__}: {e}"}


def _parse_json(raw: str) -> Optional[Dict[str, Any]]:
    raw = raw.strip()
    if raw.startswith("```"):
        raw = raw.strip("`")
        if raw.lower().startswith("json"):
            raw = raw[4:]
    try:
        return json.loads(raw)
    except Exception:
        start, end = raw.find("{"), raw.rfind("}")
        if 0 <= start < end:
            try:
                return json.loads(raw[start:end + 1])
            except Exception:
                return None
    return None
